packages/blocks/plus_minus/src/index.js

```diff
@@ -0,0 +1,1048 @@
+
+    // We need to remove args in reverse order so that it doesn't mess up
+    // as removeArg_ modifies our arrays.
+    for (let i = this.argIds_.length - 1; i >= 0; i--) {
+      this.removeArg_(this.argIds_[i]);
+    }
+
+    this.arguments_ = [];
+    this.varIds_ = [];
+    this.argumentVarModels_ = [];
+    this.argIds_ = [];
+
+    const length = varIds.length;
+    for (let i = 0; i < length; i++) {
+      this.addArg_(names[i], varIds[i]);
+    }
+
+    Blockly.Procedures.mutateCallers(this);
+  },
+
+  /**
+   * Callback for the plus image. Adds an argument to the block and mutates
+   * callers to match.
+   */
+  plus: function() {
+    this.addArg_();
+    Blockly.Procedures.mutateCallers(this);
+  },
+
+  /**
+   * Callback for the minus image. Removes the argument associated with the
+   * given argument ID and mutates the callers to match.
+   * @param {string} argId The argId of the argument to remove.
+   * @this Blockly.Block
+   */
+  minus: function(argId) {
+    if (!this.argIds_.length) {
+      return;
+    }
+    this.removeArg_(argId);
+    Blockly.Procedures.mutateCallers(this);
+  },
+
+  /**
+   * Adds an argument to the block and updates the block's parallel tracking
+   * tracking arrays as appropriate.
+   * @param {string} opt_name An optional name for the argument.
+   * @param {string} opt_varId An optional variable ID for the argument.
+   * @this Blockly.Block
+   * @private
+   */
+  addArg_: function(opt_name, opt_varId) {
+    if (!this.arguments_.length) {
+      const withField = new Blockly.FieldLabel(
+          Blockly.Msg['PROCEDURES_BEFORE_PARAMS']);
+      this.getInput('TOP')
+          .appendField(withField, 'WITH');
+    }
+
+    const name = opt_name || Blockly.Variables.generateUniqueNameFromOptions(
+        Blockly.Procedures.DEFAULT_ARG, this.arguments_);
+    const variable = Blockly.Variables.getOrCreateVariablePackage(
+        this.workspace, opt_varId, name, '');
+    const argId = Blockly.utils.genUid();
+
+    this.addVarInput_(name, argId);
+    if (this.getInput('STACK')) {
+      this.moveInputBefore(argId, 'STACK');
+    } else {
+      this.moveInputBefore(argId, 'RETURN');
+    }
+
+    this.arguments_.push(name);
+    this.varIds_.push(variable.getId());
+    this.argumentVarModels_.push(variable);
+    this.argIds_.push(argId);
+  },
+
+  /**
+   * Removes the argument associated with the given argument ID from the block.
+   * @param {string} argId An ID used to track arguments on the block.
+   * @private
+   */
+  removeArg_: function(argId) {
+    if (!this.getInput(argId)) {
+      return;
+    }
+    // TODO: My life would be so much happier if this returned a boolean instead
+    //  of throwing an error. I'm saying at least 15% more joy.
+    this.removeInput(argId);
+    if (this.arguments_.length == 1) { // Becoming argumentless.
+      this.getInput('TOP').removeField('WITH');
+    }
+
+    const index = this.argIds_.indexOf(argId);
```

Comment:
There are four arrays that you're tracking together and keeping in sync.  Instead, use a single array that contains objects, each of which has `argument`, `varId`, `varmodel`, `argId`

Reply:
Am I allowed to use .map() and .find() and things?

Reply:
Yes, you are!  (In this repo)